File: app/src/main/java/com/elementary/tasks/core/controller/EventManager.java
package com.elementary.tasks.core.controller;

import android.content.Context;

import com.elementary.tasks.ReminderApp;
import com.elementary.tasks.core.app_widgets.UpdatesHelper;
import com.elementary.tasks.core.data.AppDb;
import com.elementary.tasks.core.data.models.Reminder;
import com.elementary.tasks.core.services.PermanentReminderReceiver;
import com.elementary.tasks.core.utils.Notifier;
import com.elementary.tasks.core.utils.Prefs;

import javax.inject.Inject;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public abstract class EventManager implements EventControl {

    private Reminder mReminder;
    @Inject
    private Context mContext;
    @Inject
    private AppDb mDb;

    public Context getContext() {
        return mContext;
    }

    public Reminder getReminder() {
        return mReminder;
    }

    protected AppDb getDb() {
        return mDb;
    }

    public EventManager(Reminder reminder) {
        ReminderApp.getAppComponent().inject(this);
        this.mReminder = reminder;
    }

    protected void save() {
        mDb.reminderDao().insert(mReminder);
        UpdatesHelper.getInstance(getContext()).updateWidget();
        if (Prefs.getInstance(getContext()).isSbNotificationEnabled()) {
            Notifier.updateReminderPermanent(getContext(), PermanentReminderReceiver.ACTION_SHOW);
        }
    }
}

File: app/src/main/java/com/elementary/tasks/core/controller/EventManager.kt
package com.elementary.tasks.core.controller

import com.github.naz013.appwidgets.AppWidgetUpdater
import com.github.naz013.feature.common.coroutine.invokeSuspend
import com.elementary.tasks.core.utils.Notifier
import com.elementary.tasks.core.utils.params.Prefs
import com.github.naz013.domain.Reminder
import com.github.naz013.repository.ReminderRepository

abstract class EventManager(
  protected val reminder: Reminder,
  private val reminderRepository: ReminderRepository,
  protected val prefs: Prefs,
  protected val notifier: Notifier,
  protected val appWidgetUpdater: AppWidgetUpdater
) : EventControl {

  protected fun save() {
    invokeSuspend { reminderRepository.save(reminder) }
    appWidgetUpdater.updateAllWidgets()
    if (prefs.isSbNotificationEnabled) {
      notifier.sendShowReminderPermanent()
    }
  }

  protected fun remove() {
    invokeSuspend { reminderRepository.delete(reminder.uuId) }
    appWidgetUpdater.updateAllWidgets()
    if (prefs.isSbNotificationEnabled) {
      notifier.sendShowReminderPermanent()
    }
  }
}
